File: src/unifyweaver/targets/python_runtime/crawler.py
from collections import deque
import sys
try:
    from lxml import etree
except ImportError:
    sys.stderr.write("Error: lxml required for crawler\n")
    sys.exit(1)

from .utils import get_local

class PtCrawler:
    def __init__(self, importer, embedder=None):
        self.importer = importer
        self.embedder = embedder
        self.seen = set()

    def crawl(self, seed_ids, fetch_func, max_depth=5, embed_content=True):
        frontier = deque(seed_ids)
        self.seen.update(seed_ids)
        depth = 0
        
        while frontier and depth < max_depth:
            next_batch = []
            print(f"Depth {depth}: Processing {len(frontier)} items", file=sys.stderr)
            
            while frontier:
                obj_id = frontier.popleft()
                
                try:
                    xml_stream = fetch_func(obj_id)
                    if not xml_stream:
                        continue
                        
                    self._process_stream(xml_stream, next_batch, embed_content=embed_content)
                except Exception as e:
                    print(f"Error processing {obj_id}: {e}", file=sys.stderr)
            
            for kid in next_batch:
                frontier.append(kid)
            depth += 1

    def _process_stream(self, xml_stream, next_batch, embed_content=True):
        context = etree.iterparse(xml_stream, events=('end',), recover=True)
        count = 0
        limit = 50 # Safety limit for testing
        
        for event, elem in context:
            count += 1
            if count > limit:
                print(f"DEBUG: Reached limit of {limit} items. Stopping stream processing.", file=sys.stderr)
                break
                
            # Simple flattening (similar to read_xml_lxml)
            data = {}
            # Root element attributes (global keys for backward compatibility)
            for k, v in elem.attrib.items():
                local_k = k.split('}')[-1] if '}' in k else k
                data['@' + local_k] = v

            tag = elem.tag.split('}')[-1] # Local name

            obj_id = data.get('@id') or data.get('@rdf:about') or data.get('@about')

            # Flatten children
            for child in elem:
                child_tag = child.tag.split('}')[-1]
                # Child text content
                if not len(child) and child.text:
                    data[child_tag] = child.text.strip()

                # Child element attributes (element-scoped to prevent conflicts)
                for attr_name, attr_val in child.attrib.items():
                    local_attr = attr_name.split('}')[-1] if '}' in attr_name else attr_name
                    # Element-scoped: e.g., "seeAlso@resource" vs "parentTree@resource"
                    scoped_key = child_tag + '@' + local_attr
                    data[scoped_key] = attr_val
                    # Also store with global key for backward compatibility
                    data['@' + local_attr] = attr_val

                # Link extraction
                # Look for rdf:resource attribute
                resource = None
                for k, v in child.attrib.items():
                    # Check for }resource or just resource
                    if k.endswith('}resource') or k == 'resource':
                        resource = v
                        break

                if resource and obj_id:
                    # Store link: source=obj_id, target=resource
                    # This captures parentTree (Child->Parent) and seeAlso
                    self.importer.upsert_link(obj_id, resource)
            
            # If no ID, maybe generate one or skip? C# PtMapper extracts IDs.
            # For now, assume @id exists or skip.
            if not obj_id:
                # Check if it's a link
                pass
            else:
                self.importer.upsert_object(obj_id, tag, data)
                
                # Embeddings
                if embed_content and self.embedder:
                    text = get_local(data, 'title') or get_local(data, 'about') or data.get('text') or ""
                    if text:
                        # print(f"DEBUG: Embedding {obj_id}: {text[:50]}...", file=sys.stderr)
                        vec = self.embedder.get_embedding(text)
                        self.importer.upsert_embedding(obj_id, vec)
                
                # Links (Children)
                # If children are in data? No, children are sub-elements.
                # In lxml iterparse 'end', children are already processed?
                # Flattening logic usually handles children.
                # We need to extract links.
                # Assume specific link logic (e.g. pt:seeAlso) or generic.
                pass
            
                elem.clear()
                while elem.getprevious() is not None:
                    parent = elem.getparent()
                    if parent is not None:
                        del parent[0]
                    else:
                        break

    def process_fragments_from_stdin(self):
        """Process null-delimited XML fragments from stdin.

        This enables AWK-based ingestion where AWK filters and extracts fragments.
        Usage: awk -f extract_fragments.awk input.rdf | python_crawler
        """
        return self.process_fragments(sys.stdin.buffer)

    def process_fragments(self, reader):
        """Process null-delimited XML fragments from a binary reader.

        Each fragment is a complete XML element separated by null bytes (\\0).

        Args:
            reader: A binary file-like object (e.g., sys.stdin.buffer, open(file, 'rb'))
        """
        fragment = bytearray()
        count = 0

        # Read byte by byte, splitting on null delimiter
        while True:
            byte = reader.read(1)
            if not byte:
                # EOF - process final fragment if exists
                if fragment:
                    try:
                        self._process_fragment(bytes(fragment))
                        count += 1
                    except Exception as e:
                        print(f"Error processing fragment: {e}", file=sys.stderr)
                break

            if byte == b'\x00':
                # Found null delimiter - process accumulated fragment
                if fragment:
                    try:
                        self._process_fragment(bytes(fragment))
                        count += 1

                        if count % 100 == 0:
                            print(f"Processed {count} fragments...", file=sys.stderr)
                    except Exception as e:
                        print(f"Error processing fragment: {e}", file=sys.stderr)

                    fragment.clear()
            else:
                fragment.extend(byte)

        print(f"✓ Processed {count} total fragments", file=sys.stderr)

    def _process_fragment(self, fragment):
        """Process a single XML fragment.

        Args:
            fragment: Bytes containing a complete XML element
        """
        # Use recover=True to handle fragments without namespace declarations
        parser = etree.XMLParser(recover=True)
        try:
            root = etree.fromstring(fragment, parser=parser)
        except etree.XMLSyntaxError as e:
            raise ValueError(f"Invalid XML fragment: {e}")

        # Extract data similar to _process_stream
        data = {}

        # Root element attributes
        for k, v in root.attrib.items():
            local_k = k.split('}')[-1] if '}' in k else k
            data['@' + local_k] = v

        tag = root.tag.split('}')[-1]  # Local name
        obj_id = data.get('@id') or data.get('@rdf:about') or data.get('@about')

        # Flatten children
        for child in root:
            child_tag = child.tag.split('}')[-1]

            # Child text content
            if not len(child) and child.text:
                data[child_tag] = child.text.strip()

            # Child element attributes (element-scoped to prevent conflicts)
            for attr_name, attr_val in child.attrib.items():
                local_attr = attr_name.split('}')[-1] if '}' in attr_name else attr_name
                # Element-scoped: e.g., "seeAlso@resource" vs "parentTree@resource"
                scoped_key = child_tag + '@' + local_attr
                data[scoped_key] = attr_val
                # Also store with global key for backward compatibility
                data['@' + local_attr] = attr_val

            # Link extraction
            resource = None
            for k, v in child.attrib.items():
                if k.endswith('}resource') or k == 'resource':
                    resource = v
                    break

            if resource and obj_id:
                self.importer.upsert_link(obj_id, resource)

        # Store object if it has an ID
        if obj_id:
            self.importer.upsert_object(obj_id, tag, data)

            # Generate embedding if embedder is available
            if self.embedder:
                text = get_local(data, 'title') or get_local(data, 'about') or data.get('text') or ""
                if text:
                    vec = self.embedder.get_embedding(text)
                    self.importer.upsert_embedding(obj_id, vec)

File: src/unifyweaver/targets/rust_runtime/crawler.rs
use quick_xml::events::Event;
use quick_xml::reader::Reader;
use serde_json::{Map, Value};
use std::fs::File;
use std::io::{self, BufReader, Read};
use crate::importer::PtImporter;
use crate::embedding::EmbeddingProvider;

pub struct PtCrawler {
    importer: PtImporter,
    embedder: EmbeddingProvider,
}

impl PtCrawler {
    pub fn new(importer: PtImporter, embedder: EmbeddingProvider) -> Self {
        Self { importer, embedder }
    }

    pub fn crawl(&self, seeds: &[String], _max_depth: usize) -> Result<(), Box<dyn std::error::Error>> {
        for seed in seeds {
            // Assume seed is a file path for now
            println!("Processing {}", seed);
            self.process_file(seed)?;
        }
        Ok(())
    }

    fn process_file(&self, path: &str) -> Result<(), Box<dyn std::error::Error>> {
        let file = File::open(path)?;
        let file_len = file.metadata()?.len();
        let mut reader = Reader::from_reader(BufReader::new(file));
        // Note: trim_text config removed for quick-xml 0.31+ compatibility

        let mut buf = Vec::new();
        let mut current_obj: Option<Map<String, Value>> = None;
        let mut current_tag = String::new();

        loop {
            match reader.read_event_into(&mut buf) {
                Ok(Event::Start(e)) => {
                    let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                    // Flatten: Only care about leaf nodes or direct children of root?
                    // Strategy: Every element with an ID is an object.
                    // Attributes are props. Children text is props.
                    
                    let mut obj = Map::new();
                    let mut obj_id = String::new();
                    
                    // Extract attributes
                    for attr in e.attributes() {
                        let attr = attr?;
                        let key = String::from_utf8_lossy(attr.key.as_ref());
                        let val = String::from_utf8_lossy(&attr.value);
                        
                        if key == "rdf:about" || key == "id" || key.ends_with("about") {
                            obj_id = val.to_string();
                        }
                        
                        // Link extraction
                        if key == "rdf:resource" || key.ends_with("resource") {
                            // Store link logic handled later or here?
                            // If this is a child element, the parent ID is needed.
                            // Current simplistic approach: Flat stream.
                        }
                        
                        obj.insert(format!("@{}", key), Value::String(val.to_string()));
                    }
                    
                    // Only track top-level objects that have an ID
                    // This avoids nested elements overwriting the parent
                    if !obj_id.is_empty() && current_obj.is_none() {
                        current_obj = Some(obj);
                        current_tag = tag;
                    }
                }
                Ok(Event::Text(e)) => {
                    if let Some(ref mut obj) = current_obj {
                        let text = std::str::from_utf8(e.as_ref()).unwrap_or("").to_string();
                        if !text.is_empty() {
                            obj.insert("text".to_string(), Value::String(text));
                        }
                    }
                }
                Ok(Event::End(e)) => {
                    let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                    if let Some(obj) = current_obj.take() {
                        if tag == current_tag {
                            // Object complete
                            if let Some(Value::String(id)) = obj.get("@rdf:about").or(obj.get("@id")) {
                                self.importer.upsert_object(id, &tag, &Value::Object(obj.clone()))?;

                                // Generate Embedding
                                let text = get_local(&obj, "title")
                                    .or_else(|| get_local(&obj, "about"))
                                    .or_else(|| obj.get("text"))
                                    .and_then(|v| v.as_str());

                                if let Some(text) = text {
                                    if let Ok(vec) = self.embedder.get_embedding(text) {
                                        self.importer.upsert_embedding(id, &vec)?;
                                    }
                                }
                            }
                        } else {
                            // Put it back if it's not the matching end tag (handles nested elements)
                            current_obj = Some(obj);
                        }
                    }
                }
                Ok(Event::Eof) => break,
                Err(e) => return Err(Box::new(e)),
                _ => (),
            }
            buf.clear();
        }
        Ok(())
    }

    /// Process null-delimited XML fragments from stdin
    /// This enables AWK-based ingestion where AWK filters and extracts fragments
    /// Usage: awk -f extract_fragments.sh input.rdf | rust_crawler
    pub fn process_fragments_from_stdin(&self) -> Result<(), Box<dyn std::error::Error>> {
        self.process_fragments(&mut io::stdin())
    }

    /// Process null-delimited XML fragments from a reader
    /// Each fragment is a complete XML element separated by null bytes (\0)
    pub fn process_fragments<R: Read>(&self, reader: &mut R) -> Result<(), Box<dyn std::error::Error>> {
        let mut fragment = Vec::new();
        let mut count = 0;

        // Read byte by byte, splitting on null delimiter
        for byte_result in reader.bytes() {
            let byte = byte_result?;

            if byte == 0 {
                // Found null delimiter - process accumulated fragment
                if !fragment.is_empty() {
                    if let Err(e) = self.process_fragment(&fragment) {
                        eprintln!("Error processing fragment: {}", e);
                    }
                    count += 1;

                    if count % 100 == 0 {
                        eprintln!("Processed {} fragments...", count);
                    }

                    fragment.clear();
                }
            } else {
                fragment.push(byte);
            }
        }

        // Process final fragment if exists (no trailing null)
        if !fragment.is_empty() {
            if let Err(e) = self.process_fragment(&fragment) {
                eprintln!("Error processing fragment: {}", e);
            }
            count += 1;
        }

        eprintln!("✓ Processed {} total fragments", count);
        Ok(())
    }

    /// Process a single XML fragment
    fn process_fragment(&self, fragment: &[u8]) -> Result<(), Box<dyn std::error::Error>> {
        let mut reader = Reader::from_reader(fragment);
        // Note: trim_text config removed for quick-xml 0.31+ compatibility

        let mut buf = Vec::new();
        let mut obj: Option<Map<String, Value>> = None;
        let mut obj_id = String::new();
        let mut nested_element: Option<String> = None;
        let mut nested_text = String::new();

        loop {
            match reader.read_event_into(&mut buf) {
                Ok(Event::Start(e)) => {
                    let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();
                    let mut current_obj = Map::new();

                    // Extract attributes
                    for attr in e.attributes() {
                        let attr = attr?;
                        let key = String::from_utf8_lossy(attr.key.as_ref());
                        let val = String::from_utf8_lossy(&attr.value);

                        if key == "rdf:about" || key == "id" || key.ends_with("about") {
                            obj_id = val.to_string();
                        }

                        current_obj.insert(format!("@{}", key), Value::String(val.to_string()));
                    }

                    // Only track top-level object if it has an ID
                    if !obj_id.is_empty() && obj.is_none() {
                        obj = Some(current_obj);
                    } else if obj.is_some() {
                        // This is a nested element within our object
                        nested_element = Some(tag);
                        nested_text.clear();
                    }
                }
                Ok(Event::Text(e)) => {
                    let text = std::str::from_utf8(e.as_ref()).unwrap_or("").trim().to_string();
                    if !text.is_empty() {
                        if nested_element.is_some() {
                            // Text belongs to nested element
                            nested_text.push_str(&text);
                        } else if let Some(ref mut current_obj) = obj {
                            // Direct text content of the main object
                            current_obj.insert("text".to_string(), Value::String(text));
                        }
                    }
                }
                Ok(Event::CData(e)) => {
                    let text = std::str::from_utf8(e.as_ref()).unwrap_or("").trim().to_string();
                    if !text.is_empty() {
                        if nested_element.is_some() {
                            // CDATA belongs to nested element
                            nested_text.push_str(&text);
                        } else if let Some(ref mut current_obj) = obj {
                            // Direct CDATA of the main object
                            current_obj.insert("text".to_string(), Value::String(text));
                        }
                    }
                }
                Ok(Event::End(e)) => {
                    let tag = String::from_utf8_lossy(e.name().as_ref()).to_string();

                    // Check if this is the end of a nested element
                    if let Some(ref nested_tag) = nested_element {
                        if *nested_tag == tag {
                            // Add nested element's text to parent object
                            if let Some(ref mut current_obj) = obj {
                                if !nested_text.is_empty() {
                                    current_obj.insert(tag.clone(), Value::String(nested_text.clone()));
                                }
                            }
                            nested_element = None;
                            nested_text.clear();
                            buf.clear();
                            continue;
                        }
                    }

                    // This is the end of the main object
                    if let Some(current_obj) = obj.take() {
                        // Extract ID from attributes
                        let id = current_obj.get("@rdf:about")
                            .or(current_obj.get("@id"))
                            .or(current_obj.get("@about"))
                            .and_then(|v| v.as_str())
                            .unwrap_or("");

                        if !id.is_empty() {
                            // Get tag for object type (could extract from fragment)
                            let tag = "unknown";
                            self.importer.upsert_object(id, tag, &Value::Object(current_obj.clone()))?;

                            // Generate embedding if we have text
                            let text = get_local(&current_obj, "title")
                                .or_else(|| get_local(&current_obj, "about"))
                                .or_else(|| current_obj.get("text"))
                                .and_then(|v| v.as_str());

                            if let Some(text) = text {
                                if let Ok(vec) = self.embedder.get_embedding(text) {
                                    self.importer.upsert_embedding(id, &vec)?;
                                }
                            }
                        }
                    }
                }
                Ok(Event::Eof) => break,
                Err(e) => return Err(Box::new(e)),
                _ => (),
            }
            buf.clear();
        }

        Ok(())
    }
}

/// Helper function to get a value from a JSON map ignoring namespace prefixes
/// Tries: exact match, @prefix, and any namespace:localname pattern
fn get_local<'a>(map: &'a serde_json::Map<String, Value>, local_name: &str) -> Option<&'a Value> {
    // Try exact match
    if let Some(v) = map.get(local_name) {
        return Some(v);
    }

    // Try with @ prefix
    let at_key = format!("@{}", local_name);
    if let Some(v) = map.get(&at_key) {
        return Some(v);
    }

    // Try with any namespace prefix (e.g., "dcterms:title", "@dcterms:title")
    let suffix = format!(":{}", local_name);
    for (key, value) in map.iter() {
        if key.ends_with(&suffix) {
            return Some(value);
        }
    }

    None
}
